### PR TITLE
Modularize status tracking for pools of components in the power distributing actor

### PR DESCRIPTION
**Pull request by shsms**
The new design works as follows:

- reusable `ComponentPoolStatusTracker` that takes individual `ComponentStatus` and generates `ComponentPoolStatus` objects.
- the individual `ComponentStatus` need to be generated by code specific to the component category, because status is calculated differently for different component categories.  This code specific for a component category has to be a class that implements the abstract `ComponentStatusTracker` class.
- The `ComponentPoolStatusTracker` accepts implementations of `ComponentStatusTracker`, like the `BatteryStatusTracker` (which looks at the status of corresponding inverters also), and but is otherwise generic.

This PR takes us part of the way towards getting the PowerDistributingActor easily usable for other component types.

It also simplifies some implementations and replaces the use of `MergeNamed` with `Merge` from the channels repo.

### CODE REVIEW

**Comment by shsms** [comment]
No externally visible changes, so no release notes.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:2]:
```suggestion
# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
```

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py]: batteries?

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py]: I think this is mixing present and past for succeed and failed, respectively. If this comment applies, then argument names need to be updated too

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py:57]: what about making it `kw_only=True` too?

File: src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py:177]: The battery should remain in the documentation, right?

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py]: I think here it should be `component` instead of `battery`

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py]: I think the component status interface shouldn't talk about batteries

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py:15]: can this dataclass be `frozen=True, kw_only=True` or there is a reason to not do it?

**Review by daniel-zullo-frequenz — COMMENTED** [review]
I've done my first review round and made a few suggestions to check for. I'll have another look after those suggestions are discussed/addressed.

File: src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py
Comment (llucax) [inline src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py:168]: Not a big deal but any reason to mix the `battery_id` → `component_id` rename with the `_battery_status.py` → `_battery_status_tracker.py`? They seem pretty unrelated.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (llucax) [inline src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:26]: I guess not for this PR, but can't we take the opportunity and include the migration to `BackgroudService` in this big refactoring. These classes handling multiple tasks and having a `stop()` and `join()` look like a perfect candidate for `BackgorundService`.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py:23]: Also probably `collections.abc.Set` so the user can't mutate it either.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py:66]: I'm not convinced about this change. Can't you just use `select()` if you need to know the source?  I guess you'll still need a mapping between receiver and `component_id`, but otherwise this is mostly doing the same, but you need to create one extra class that has a name clash with the `Enum`, so now you need to add the `Enum` suffix to the enum, which is weird and we never do (it could be like adding a `Class` suffix to classes.

If you think it is better to stick with `merge` and this mapping wrapper, then I suggest to find some other name to avoid the `Enum` suffix, maybe `ComponentStatus` and `ComponentState`, or `CurrentComponentState` to avoid confusion between status and state?

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py]: Since you are using `set`s internally here, wouldn't it also make more sense to use `collections.abc.Set`? It seems like it could be useful for users to be able to do `if id in result.succeeded` for example.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:2]: This is actually a renaming, not a new file: https://github.com/frequenz-floss/frequenz-sdk-python/pull/779/commits/278b0b92475dc3d42566560a5d4493095f7bdd26#diff-40a429480bdbbab4e414638f059fa7202095acd519c72353f9ad2a762e75cad7

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [inline src/frequenz/sdk/actor/power_distributing/_component_status.py:79]: This one too could be a `BackgroundService`.

Now I wonder if we should maybe have a purely abstract `BackgroundService` class (or even protocol), and then a `BaseBackgroundService` to use as a convenience default implementation, as in here you are only defining an interface. I think `BackgroundService` should also be only an interface, as we have background services that are managing other background services instead of tasks directly, and I ended up mixing up things a bit for the sake of convenience (interface/protocol vs base mixin implementation).

But in any case for now `BackgroundService` could be used too, it will just add an unnecessary `_tasks` member to the class.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (llucax) [inline src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:93]: It might be better to return a `Receiver` here as `Merge` is becoming private.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py]: I'll fix this particular case, but I'm sure there are other instances as well.  This is old code, and this PR has a limited scope.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:57]: the two fields have different types.

File: src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py:177]: yes, I guess it doesn't matter.  had to change the field to make it fit into a generic interface.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: true, will fix.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: fixing

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:15]: probably, but out of scope for this PR.

File: src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py:168]: oops, guess I mixed up some commits.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:26]: maybe, but right now I'm just doing the bare minimum so I can unblock *pools.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:23]: there's a lot of non-bug style issues with the code base.  This PR is just about renaming some things and modularizing some things.  I did improve some things, but don't want to get too deep into fixing other style issues at the moment in internal code.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:66]: We need the component id from a variable number of components, so implementation-wise I think this is the best approach.

I think they are both status or both state, but one of them is tied to a specific component while the other is not.  I agree that `Enum` is not perfect, but I couldn't come up with a better one.

I would like the class with the id to be called `ComponentStatus`, to form a pattern with the `ComponentPoolStatus`, `ComponentStatusTracker`, `ComponentPoolStatusTracker`.  I guess we just need an alternative for the enum.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: it wasn't me,  it was already like that.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:93]: So we stop calling `stop` on it in tests, etc, right?

File: src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_battery_status_tracker.py:168]: but I seem to have mentioned this in the commit message, so looks like it wasn't a mixup.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py]: Thanks!

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:66]: > We need the component id from a variable number of components, so implementation-wise I think this is the best approach.

just to contextualize this, this is the case if we want to limit the changes to this part of the code base.  it would be possible to avoid the merge if we rewrite the `ComponentPoolStatusTracker`, but that won't be a trivial change.

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py]: fixed in: https://github.com/frequenz-floss/frequenz-sdk-python/commit/61765a81

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py]: done here: https://github.com/frequenz-floss/frequenz-sdk-python/commit/61765a81

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: https://github.com/frequenz-floss/frequenz-sdk-python/commit/61765a81

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: https://github.com/frequenz-floss/frequenz-sdk-python/commit/61765a81

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: OK, but you are changing members names already, so it might be good to add to the bag of refactoring (it is fine if it's not in this PR but it looks like a trivial change).

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:26]: I agree for the stuff that was already there and was just renamed/moved, but for new classes I think we should start using `BackgroundService`, it makes no sense to keep piling up technical debt.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:66]: > I would like the class with the id to be called `ComponentStatus`, to form a pattern with the `ComponentPoolStatus`, `ComponentStatusTracker`, `ComponentPoolStatusTracker`. I guess we just need an alternative for the enum.

I see. I think, besides being a bit ambiguous, `ComponentState` for the enum makes sense. Status and state is not exactly the same. For example a state machine is never called status machine, and "WORKING", etc. look like states in a state machine, is like an abstract concept, while status conveys more the meaning of something that's happening at a particular point in time. For example a status report wouldn't be called state report, and it talks about one point in time. `ComponentStatus` says this particular component is like this right now.

Anyway, I agree it has the downside of being a bit ambiguous. Not a perfect solution so I would leave it to you to decide.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:79]: Like this class is new, right?

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:93]: Mmm, damn. Houston, we have a problem. Another reason to use `select()` instead. :-P

But seriously, it looks like we can't make `_Merge` private in channels then.

**Comment by llucax** [comment]
I just went through your replies and I'm fine with leaving things as they are. The only thing I really wouldn't like is to keep adding classes that should inherit from `BackgroundService` without doing so. If all the classes here are really just renames/moves, then I'm also fine with keeping it like it is in this PR and will approve.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:66]: I was thinking that another option, and in general for all enums, it might make sense to name them in plural, as it is a collection of names or stuff, specially in Python that they are even iterable.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: I already did a few trivial changes, there's just so many, but ok, I will do this one too.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:79]: nope, it is just an abc for an existing `BatteryStatusTracker` class.

But I can look into this again as I start working on the EVChargerStatusTracker, which will be new.  And migrate this too at that point.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:23]: I tried to do this also, but it doesn't work because we don't have some methods like `intersection`, etc. on the abstract types.

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py]: Done here: https://github.com/frequenz-floss/frequenz-sdk-python/commit/00c48d9e

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:119]: We might want to a short delay between restarts. This might be also out of scope for this PR but worth mentioning it

**Review by daniel-zullo-frequenz — COMMENTED** [review]
I don't have any further comments within the scope of this PR, LGTM
Though Luca's comments need to be resolve so I'll leave it to him to approve it

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:23]: Really? That's lame :scream:

File: tests/actor/test_battery_status.py
Comment (llucax) [inline tests/actor/test_battery_status.py:541]: Ha! This change uncovered a bug-ish (I mean, it was not a bug because it was returning an iterable, but still).

**Review by llucax — APPROVED** [review]

File: src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:119]: Wow, I hadn't seen this.  I think this would never happen, because we don't close any of our channels and that's the only source of an exception here.

I think I'll just remove this exception handling, causing the actor to restart fully instead.

Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_pool_status_tracker.py:119]: Will do in the next PR.

File: tests/actor/test_battery_status.py
Comment (shsms) [reply tests/actor/test_battery_status.py:541]: yes, the great traps in python. :-/

File: src/frequenz/sdk/actor/power_distributing/_component_status.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_status.py:23]: it turns out this is not a read-only class.  when we get news about a single battery, we update the corresponding list.

**Pull request merged**
